### PR TITLE
Implement Instance Normalization

### PR DESCRIPTION
**Pull request by tushushu**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs
https://github.com/tracel-ai/burn/issues/1249

### Changes
Implementing the Instance Normalization based on `GroupNorm`.

### Testing
It's tested by the `instance_norm_forward_affine_false` and `instance_norm_forward_affine_true` functions.

### CODE REVIEW

**Comment by tushushu** [comment]
The `run-checks.sh` successfully finished with ENV arguments in ('no-std', 'typos', 'examples'), but failed with ENV arg = 'std' because torch compilation error below, and I suppose it is not related to my PR.
```
error: failed to run custom build command for `torch-sys v0.15.0`

Caused by:
  process didn't exit successfully: `/burn/target/debug/build/torch-sys-f941c31579c5fe6a/build-script-build` (exit status: 1)
  --- stdout
```

**Comment by Arjun31415** [comment]
Maybe squash the Commits

**Comment by tushushu** [comment]
> Maybe squash the Commits
I usually use squash merge instead, does it make sense or in this community everyone has to use squash commit?

**Comment by antimora** [comment]
> > Maybe squash the Commits
> 
> I usually use squash merge instead, does it make sense or in this community everyone has to use squash commit?

The commits are squashed during PR merge. Do not worry about this now.

**Comment by antimora** [comment]
I approved the CI checks (first time contributors need to be approved, later it's automatic). So we will find out if tch backend is working.

**Review by antimora — COMMENTED** [review]
Can we add ✅ for Burn support in ONNX?

https://github.com/tracel-ai/burn/blob/main/burn-import/SUPPORTED-ONNX-OPS.md?plain=1#L87

This is just documentation update no need to implement ONNX OP (unless you want to take up in a new PR).

**Comment by tushushu** [comment]
> https://github.com/tracel-ai/burn/blob/main/burn-import/SUPPORTED-ONNX-OPS.md?plain=1#L87

I've added ✅ for both InstanceNorm and GroupNorm in the doc, could you review the PR again?

File: burn-core/src/nn/norm/instance.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    /// Initialize a new [instance norm](InstanceNorm) module with a [record](InstanceNormRecord).
+    pub fn init_with<B: Backend>(&self, record: InstanceNormRecord<B>) -> InstanceNorm<B> {
+        InstanceNorm {
+            group_norm: self.into().init_with(record.group_norm),
+        }
+    }
+
+    fn into(&self) -> GroupNormConfig {
+        GroupNormConfig {
+            num_groups: self.num_channels,
```
Comment (nathanielsimard) [inline burn-core/src/nn/norm/instance.rs:47]: I would add a comment as to why instance norm is equivalent to group norm with this line.

**Review by nathanielsimard — COMMENTED** [review]
LGTM, but tagging @louisfd for review.

File: burn-core/src/nn/norm/instance.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            num_channels: self.num_channels,
+            epsilon: self.epsilon,
+            affine: self.affine,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (tushushu) [reply burn-core/src/nn/norm/instance.rs]: Sounds good to me.

File: burn-core/src/nn/norm/instance.rs (hunk 1 above)
Comment (tushushu) [reply burn-core/src/nn/norm/instance.rs:49]: I've added a comment to explain.

**Review by louisfd — APPROVED** [review]
LGTM

**Review by antimora — APPROVED** [review]
LGTM

**Review by nathanielsimard — APPROVED** [review]
I think the only missing part is merging main and fixing the conflicts, then we'll merge! Thanks a lot.

**Pull request merged**